File: lv1_module2/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <chrono>
#include <cmath>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rcl_interfaces/msg/set_parameters_result.hpp"
#include "std_msgs/msg/float32.hpp"
#include "turtlesim/msg/pose.hpp"

using namespace std::chrono_literals;

class DistancePublisher : public rclcpp::Node
{
public:
  DistancePublisher()
  : Node("cpp_distance_publisher")
  {
    publish_rate_ = declare_parameter<double>("publish_rate", 10.0);
    if (publish_rate_ <= 0.0) {
      RCLCPP_WARN(get_logger(), "publish_rate must be positive; using 10.0 Hz");
      publish_rate_ = 10.0;
    }

    const auto pose_topic = declare_parameter<std::string>("pose_topic", "/turtle1/pose");
    const auto distance_topic =
      declare_parameter<std::string>("distance_topic", "/turtle_distance");

    publisher_ = create_publisher<std_msgs::msg::Float32>(distance_topic, 10);
    subscription_ = create_subscription<turtlesim::msg::Pose>(
      pose_topic,
      10,
      [this](const turtlesim::msg::Pose::SharedPtr message) {
        latest_x_ = message->x;
        latest_y_ = message->y;
        pose_received_ = true;
      });

    create_publish_timer();

    parameter_callback_handle_ = add_on_set_parameters_callback(
      std::bind(&DistancePublisher::on_parameters, this, std::placeholders::_1));

    RCLCPP_INFO(
      get_logger(), "C++ distance publisher: %s -> %s at %.1f Hz",
      pose_topic.c_str(), distance_topic.c_str(), publish_rate_);
  }

private:
  void create_publish_timer()
  {
    const auto period = std::chrono::duration<double>(1.0 / publish_rate_);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(period),
      [this]() {
        if (!pose_received_) {
          return;
        }
        std_msgs::msg::Float32 message;
        message.data = static_cast<float>(std::hypot(latest_x_, latest_y_));
        publisher_->publish(message);
      });
  }

  rcl_interfaces::msg::SetParametersResult on_parameters(
    const std::vector<rclcpp::Parameter> & parameters)
  {
    rcl_interfaces::msg::SetParametersResult result;
    result.successful = true;

    for (const auto & parameter : parameters) {
      if (parameter.get_name() == "publish_rate") {
        const double requested_rate = parameter.as_double();
        if (requested_rate <= 0.0) {
          result.successful = false;
          result.reason = "publish_rate must be greater than zero";
          RCLCPP_WARN(get_logger(), "%s", result.reason.c_str());
          return result;
        }

        publish_rate_ = requested_rate;
        timer_->cancel();
        create_publish_timer();
        RCLCPP_INFO(get_logger(), "publish_rate changed to %.1f Hz", publish_rate_);
      }
    }
    return result;
  }

  double publish_rate_{10.0};
  double latest_x_{0.0};
  double latest_y_{0.0};
  bool pose_received_{false};
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr publisher_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr timer_;
  OnSetParametersCallbackHandle::SharedPtr parameter_callback_handle_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<DistancePublisher>();
  rclcpp::spin(node);
  node.reset();
  rclcpp::shutdown();
  return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
"""Publish the turtle's distance from the origin at a configurable rate."""

import math

import rclpy
from rcl_interfaces.msg import SetParametersResult
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import Float32
from turtlesim.msg import Pose

from turtle_py.qos_utils import make_qos_profile


class DistancePublisher(Node):
    def __init__(self) -> None:
        super().__init__('distance_publisher')

        self.declare_parameter('publish_rate', 10.0)
        self.declare_parameter('pose_topic', '/turtle1/pose')
        self.declare_parameter('distance_topic', '/turtle_distance')
        self.declare_parameter('pose_timeout', 1.0)
        self.declare_parameter('reliability', 'reliable')
        self.declare_parameter('durability', 'volatile')
        self.declare_parameter('qos_depth', 10)

        self._publish_rate = float(self.get_parameter('publish_rate').value)
        if self._publish_rate <= 0.0:
            self.get_logger().warning(
                'publish_rate must be greater than zero; using 10.0 Hz')
            self._publish_rate = 10.0

        pose_topic = str(self.get_parameter('pose_topic').value)
        distance_topic = str(self.get_parameter('distance_topic').value)
        self._pose_timeout = max(
            0.0, float(self.get_parameter('pose_timeout').value))
        reliability = str(self.get_parameter('reliability').value)
        durability = str(self.get_parameter('durability').value)
        qos_depth = int(self.get_parameter('qos_depth').value)

        try:
            distance_qos = make_qos_profile(reliability, durability, qos_depth)
        except ValueError as error:
            self.get_logger().warning(
                f'{error}; using reliable/volatile/depth 10')
            distance_qos = make_qos_profile()

        self._latest_pose: Pose | None = None
        self._last_pose_time = None
        self._pose_stale_warned = False
        self._publisher = self.create_publisher(Float32, distance_topic, distance_qos)
        self._subscription = self.create_subscription(
            Pose,
            pose_topic,
            self._on_pose,
            10,
        )
        self._timer = self.create_timer(1.0 / self._publish_rate, self._on_timer)
        self.add_on_set_parameters_callback(self._on_parameters)

        self.get_logger().info(
            f'distance publisher: {pose_topic} -> {distance_topic}, '
            f'{self._publish_rate:.1f} Hz, {reliability}/{durability}')

    def _on_pose(self, message: Pose) -> None:
        # 구독 콜백은 최신 자세만 저장한다.
        self._latest_pose = message
        self._last_pose_time = self.get_clock().now()
        self._pose_stale_warned = False

    def _on_timer(self) -> None:
        # 실제 발행은 타이머 콜백에서 수행한다.
        if self._latest_pose is None:
            return
        if self._pose_timeout > 0.0 and self._last_pose_time is not None:
            pose_age = (
                self.get_clock().now() - self._last_pose_time
            ).nanoseconds / 1_000_000_000.0
            if pose_age > self._pose_timeout:
                if not self._pose_stale_warned:
                    self.get_logger().warning(
                        f'no fresh pose for {pose_age:.2f} s; '
                        'suppressing /turtle_distance output')
                    self._pose_stale_warned = True
                return
        message = Float32()
        message.data = float(math.hypot(self._latest_pose.x, self._latest_pose.y))
        self._publisher.publish(message)

    def _on_parameters(self, parameters) -> SetParametersResult:
        for parameter in parameters:
            if parameter.name == 'publish_rate':
                requested_rate = float(parameter.value)
                if requested_rate <= 0.0:
                    reason = 'publish_rate must be greater than zero'
                    self.get_logger().warning(reason)
                    return SetParametersResult(successful=False, reason=reason)

                self._publish_rate = requested_rate
                self.destroy_timer(self._timer)
                self._timer = self.create_timer(
                    1.0 / self._publish_rate,
                    self._on_timer,
                )
                self.get_logger().info(
                    f'publish_rate changed to {self._publish_rate:.1f} Hz')

        return SetParametersResult(successful=True)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = DistancePublisher()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        if rclpy.ok():
            node.get_logger().info('shutdown requested; stopping cleanly')
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()
